cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityManagementService.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.capability;
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.AbstractRetryableScheduledService;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.internal.app.services.SystemProgramManagementService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Scan service for capability configurations
+ */
+public class CapabilityManagementService extends AbstractRetryableScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CapabilityManagementService.class);
+  private static final Gson GSON = new Gson();
+  private final long scheduleIntervalInMillis;
+  private final CConfiguration cConf;
+  private final CapabilityApplier capabilityApplier;
+  private final SystemProgramManagementService systemProgramManagementService;
+
+  @Inject
+  CapabilityManagementService(CConfiguration cConf, CapabilityApplier capabilityApplier,
+                              SystemProgramManagementService systemProgramManagementService) {
+    super(RetryStrategies
+            .fixDelay(cConf.getLong(Constants.Capability.DIR_SCAN_INTERVAL_MINUTES), TimeUnit.MINUTES));
+    this.cConf = cConf;
+    this.capabilityApplier = capabilityApplier;
+    this.systemProgramManagementService = systemProgramManagementService;
+    this.scheduleIntervalInMillis = TimeUnit.MINUTES
+      .toMillis(cConf.getLong(Constants.Capability.DIR_SCAN_INTERVAL_MINUTES));
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    systemProgramManagementService.start();
+  }
+
+  @Override
+  protected void doShutdown() throws Exception {
+    super.doShutdown();
+    systemProgramManagementService.stopAndWait();
+  }
+
+  @Override
+  protected long runTask() throws Exception {
+    LOG.debug("Scanning capability config directory.");
+    List<CapabilityConfig> capabilityConfigList = scanConfigDirectory();
+    //apply all the config
+    capabilityApplier.apply(capabilityConfigList);
+    return scheduleIntervalInMillis;
+  }
+
+  private List<CapabilityConfig> scanConfigDirectory() throws IOException {
+    File configDir = new File(cConf.get(Constants.Capability.CONFIG_DIR));
+    List<CapabilityConfig> allConfigs = new ArrayList<>();
```

Comment:
It is a better habit to keep naming consistent in the same class. E.g. it is called `allConfigs` here, while it is called `capabilityConfigList` from the method above. It is better to consistently calling it `capabilityConfigs` (plural of the collection type)

Reply:
Done